src/lib/hash/sha2_32/sha2_32.h

```diff
@@ -16,70 +16,96 @@ namespace Botan {
 /**
 * SHA-224
 */
-class SHA_224 final : public MDx_HashFunction {
+class SHA_224 final : public HashFunction {
+   public:
+      using digest_type = secure_vector<uint32_t>;
+
+      static constexpr MD_Endian byte_endianness = MD_Endian::Big;
+      static constexpr MD_Endian bit_endianness = MD_Endian::Big;
+      static constexpr size_t block_bytes = 64;
+      static constexpr size_t output_bytes = 28;
+      static constexpr size_t ctr_bytes = 8;
+
+      static void compress_n(digest_type& digest, std::span<const uint8_t> input, size_t blocks);
+      static void init(digest_type& digest);
+
    public:
       std::string name() const override { return "SHA-224"; }
 
-      size_t output_length() const override { return 28; }
+      size_t output_length() const override { return output_bytes; }
+
+      size_t hash_block_size() const override { return block_bytes; }
 
       std::unique_ptr<HashFunction> new_object() const override { return std::make_unique<SHA_224>(); }
 
-      std::unique_ptr<HashFunction> copy_state() const override;
+      std::unique_ptr<HashFunction> copy_state() const override { return std::make_unique<SHA_224>(*this); }
 
-      void clear() override;
+      void clear() override { m_md.clear(); }
 
       std::string provider() const override;
 
-      SHA_224() : MDx_HashFunction(64, true, true), m_digest(8) { clear(); }
-
    private:
-      void compress_n(const uint8_t[], size_t blocks) override;
-      void copy_out(uint8_t[]) override;
+      void add_data(std::span<const uint8_t> input) override { m_md.update(input); }
+
+      void final_result(std::span<uint8_t> output) override { m_md.final(output); }
```

Comment:
I don't think inlining these makes sense. For application users, they can't take advantage of the inlined nature. And it blows up code size for anything within the library that's operating on the concrete types.